hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+
+Hash tables achieve their speedy lookup times of O(1) by storing their key-value pairs into an array. They take advantage of how arrays have O(1) lookup operations when given an index.
+
+Each hash table has an _internal array_ used for storing. When a hash table inserts a key-value pair, it stores the value into the _internal array_.
+
+Hash tables use _hash functions_ to determine how to add the value into the internal array. **Hash functions** convert a given key into an index for the internal array. Hash tables store the values into that internal array using the computed index, essentially storing the entire key-value pair.
+
+Because the hash function can quickly convert a key to an array index, this operation can result in very quick lookup times.
+
+<!-- Lucidchart link  https://www.lucidchart.com/invitations/accept/5fdcf503-7d8b-4139-94d4-795bfed27883 -->
+
+![Hash Function Example](../assets/hash-tables_hash-tables_hash-function.png)
+
+<!-- Hash function image source: https://en.wikipedia.org/wiki/Hash_function -->
+
+### !callout-info
+
+## Hash Tables Implement Associative Arrays
+
+One reason why hash tables are interesting is because they implement an _abstract data type_ known as an _associative array_. An _abstract data type_ is any data structure that is theoretical, and described not by its concrete implementation, but by its behavior and how it's used. An _associative array_ is a theoretical data structure that describes collections that store key-value pairs.
+
+### !end-callout
+
+## Time Complexity
+
+For all three operations of inserting, finding, and deleting values, hash tables have an O(1) time complexity.
+
+| Action                        | Average Case | Worst Case |
+| ----------------------------- | ------------ | ---------- |
+| Finding a value by key        | O(1)         | O(n)       |
+| Removing a value by key       | O(1)         | O(n)       |
+| Inserting a new key and value | O(1)         | O(n)       |
+
+### !callout-info
+
+## We Describe Hash Table Big O by Average Case
+
+Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average-case**. The likelihood of the worst case from occurring depends on how well the **hash function** is implemented.
+
+### !end-callout
+
+## Hash Function
+
+**Hash functions** are functions responsible for mapping a given key into an integer index in the internal array.
+
+### Example Hash Function
+
+Imagine that we have a hash table which holds "name"-"job title" key-value pairs. This hash table needs a hash function to transform the "name" into an index.
+
```

Comment:
move the hash function image here
```suggestion

![Applying a hashing algorithm to several names. There are 5 steps: Key, Numeric Conversion, Numeric Value, mod by Array Length, and Index. The first key is Debbie Nguyen. An arrow indicates that we perform a numeric conversion on the name, resulting in 13 for the numeric value. Another arrow indicates the mod step (mod 4 since there are 4 indices) resulting in index 1. This will end up being a collision and is shown in fuchsia. The remaining 4 sets of data are Carlton Griffith→16→0 which will be a collision, Eddie Bowers→12→0 which is a collision with Carlton Griffith, Kaye Campbell→13→1 which is a collision with Debbie Nguyen, and Jeanette Barker→15→3 which is not a collision and is shown in yellow.](../assets/hash-tables_hash-tables_hash-function.png)  
_Fig. Mapping names to indices. Here, we used `len()` to convert each name to a number, then performed a mod by 4 to find the index in the internal array, since the internal array currently has 4 slots. Collisions can happen as a result of either the initial numeric conversion, or as a result of the squeezing forced by the modulo operation._

```